File: backend/src/controllers/projectController.ts
import { Request, Response } from 'express';
import { ProjectService } from '../services/projectService.js';
import { z } from 'zod';

const createProjectSchema = z.object({
  name: z.string().min(1, 'Project name is required').max(100, 'Project name too long'),
  description: z.string().max(500, 'Description too long').optional(),
  isPublic: z.boolean().optional(),
});

const updateProjectSchema = z.object({
  name: z.string().min(1, 'Project name is required').max(100, 'Project name too long').optional(),
  description: z.string().max(500, 'Description too long').optional(),
  isPublic: z.boolean().optional(),
});

/**
 * Project controller handles project management endpoints
 * Provides CRUD operations with proper authentication
 */
export class ProjectController {
  private projectService: ProjectService;

  constructor() {
    this.projectService = new ProjectService();
  }

  /**
   * Get all projects for the authenticated user
   * GET /api/projects
   */
  getUserProjects = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const projects = await this.projectService.getUserProjects(req.user.userId);

      res.json({
        projects,
        total: projects.length,
      });
    } catch (error) {
      console.error('Get user projects error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Get a single project by ID
   * GET /api/projects/:id
   */
  getProject = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const projectId = parseInt(req.params.id);
      if (isNaN(projectId)) {
        res.status(400).json({ error: 'Invalid project ID' });
        return;
      }

      const project = await this.projectService.getProjectById(projectId, req.user.userId);

      if (!project) {
        res.status(404).json({ error: 'Project not found' });
        return;
      }

      res.json({ project });
    } catch (error) {
      console.error('Get project error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Create a new project
   * POST /api/projects
   */
  createProject = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const validatedData = createProjectSchema.parse(req.body);

      const project = await this.projectService.createProject(req.user.userId, validatedData);

      res.status(201).json({
        message: 'Project created successfully',
        project,
      });
    } catch (error) {
      if (error instanceof z.ZodError) {
        res.status(400).json({
          error: 'Validation failed',
          details: error.errors,
        });
        return;
      }

      console.error('Create project error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Update a project
   * PUT /api/projects/:id
   */
  updateProject = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const projectId = parseInt(req.params.id);
      if (isNaN(projectId)) {
        res.status(400).json({ error: 'Invalid project ID' });
        return;
      }

      const validatedData = updateProjectSchema.parse(req.body);

      const updatedProject = await this.projectService.updateProject(
        projectId,
        req.user.userId,
        validatedData
      );

      if (!updatedProject) {
        res.status(404).json({ error: 'Project not found' });
        return;
      }

      res.json({
        message: 'Project updated successfully',
        project: updatedProject,
      });
    } catch (error) {
      if (error instanceof z.ZodError) {
        res.status(400).json({
          error: 'Validation failed',
          details: error.errors,
        });
        return;
      }

      console.error('Update project error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Delete a project
   * DELETE /api/projects/:id
   */
  deleteProject = async (req: Request, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'User not authenticated' });
        return;
      }

      const projectId = parseInt(req.params.id);
      if (isNaN(projectId)) {
        res.status(400).json({ error: 'Invalid project ID' });
        return;
      }

      const deleted = await this.projectService.deleteProject(projectId, req.user.userId);

      if (!deleted) {
        res.status(404).json({ error: 'Project not found' });
        return;
      }

      res.json({ message: 'Project deleted successfully' });
    } catch (error) {
      console.error('Delete project error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };

  /**
   * Get public projects for discovery
   * GET /api/projects/public
   */
  getPublicProjects = async (req: Request, res: Response): Promise<void> => {
    try {
      const limit = parseInt(req.query.limit as string) || 20;

      const projects = await this.projectService.getPublicProjects(limit);

      res.json({
        projects,
        total: projects.length,
      });
    } catch (error) {
      console.error('Get public projects error:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  };
}

File: backend-springboot/src/main/java/com/learningadvisor/backend/controller/ProjectController.java
package com.learningadvisor.backend.controller;

import com.learningadvisor.backend.dto.ApiResponse;
import com.learningadvisor.backend.dto.CreateProjectRequest;
import com.learningadvisor.backend.dto.ProjectDTO;
import com.learningadvisor.backend.dto.UpdateProjectRequest;
import com.learningadvisor.backend.service.ProjectService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * REST controller for project management endpoints
 * Handles CRUD operations for user projects
 */
@RestController
@RequestMapping("/api/projects")
@RequiredArgsConstructor
public class ProjectController {
    private final ProjectService projectService;

    /**
     * Get all projects for the authenticated user
     * GET /api/projects
     */
    @GetMapping
    public ResponseEntity<Map<String, Object>> getUserProjects(Authentication authentication) {
        Long userId = extractUserId(authentication);
        List<ProjectDTO> projects = projectService.getUserProjects(userId);

        Map<String, Object> response = new HashMap<>();
        response.put("projects", projects);
        response.put("total", projects.size());

        return ResponseEntity.ok(response);
    }

    /**
     * Get a single project by ID
     * GET /api/projects/{id}
     */
    @GetMapping("/{id}")
    public ResponseEntity<?> getProject(@PathVariable Long id, Authentication authentication) {
        Long userId = extractUserId(authentication);
        Optional<ProjectDTO> project = projectService.getProjectById(id, userId);

        if (project.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("error", "Project not found"));
        }

        return ResponseEntity.ok(Map.of("project", project.get()));
    }

    /**
     * Create a new project
     * POST /api/projects
     */
    @PostMapping
    public ResponseEntity<Map<String, Object>> createProject(
            @Valid @RequestBody CreateProjectRequest request,
            Authentication authentication) {
        Long userId = extractUserId(authentication);
        ProjectDTO project = projectService.createProject(userId, request);

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Project created successfully");
        response.put("project", project);

        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    /**
     * Update a project
     * PUT /api/projects/{id}
     */
    @PutMapping("/{id}")
    public ResponseEntity<?> updateProject(
            @PathVariable Long id,
            @Valid @RequestBody UpdateProjectRequest request,
            Authentication authentication) {
        Long userId = extractUserId(authentication);
        Optional<ProjectDTO> updatedProject = projectService.updateProject(id, userId, request);

        if (updatedProject.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("error", "Project not found"));
        }

        Map<String, Object> response = new HashMap<>();
        response.put("message", "Project updated successfully");
        response.put("project", updatedProject.get());

        return ResponseEntity.ok(response);
    }

    /**
     * Delete a project
     * DELETE /api/projects/{id}
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteProject(@PathVariable Long id, Authentication authentication) {
        Long userId = extractUserId(authentication);
        boolean deleted = projectService.deleteProject(id, userId);

        if (!deleted) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("error", "Project not found"));
        }

        return ResponseEntity.ok(Map.of("message", "Project deleted successfully"));
    }

    /**
     * Get public projects for discovery
     * GET /api/projects/public
     */
    @GetMapping("/public")
    public ResponseEntity<Map<String, Object>> getPublicProjects(
            @RequestParam(defaultValue = "20") int limit) {
        List<ProjectDTO> projects = projectService.getPublicProjects(limit);

        Map<String, Object> response = new HashMap<>();
        response.put("projects", projects);
        response.put("total", projects.size());

        return ResponseEntity.ok(response);
    }

    /**
     * Extract user ID from authentication
     */
    private Long extractUserId(Authentication authentication) {
        return Long.parseLong(authentication.getName());
    }
}
